File: examples/clj/test/com/rpl/agent/basic/async_agent_test.clj
(ns com.rpl.agent.basic.async-agent-test
  (:require
   [clojure.test :refer [deftest testing is]]
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest]
   [com.rpl.agent.basic.async-agent :refer [AsyncAgentModule]]))

(deftest async-agent-test
  (testing "AsyncAgent example produces expected results"
    (System/gc)
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc AsyncAgentModule {:tasks 1 :threads 1})
      (let [manager (aor/agent-manager
                     ipc
                     (rama/get-module-name AsyncAgentModule))
            agent   (aor/agent-client manager "AsyncAgent")]
        (testing "synchronous invocation produces expected result"
          (let [result (aor/agent-invoke agent "Test Task")]
            (is (= "Task 'Test Task' completed successfully" result))))

        (testing "asynchronous initiation and result produces expected result"
          (let [invoke (aor/agent-initiate agent "Async Task")
                result (aor/agent-result agent invoke)]
            (is (= "Task 'Async Task' completed successfully" result))))))))

File: examples/java/src/test/java/com/rpl/agent/basic/AsyncAgentTest.java
package com.rpl.agent.basic;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentInvoke;
import com.rpl.agentorama.AgentManager;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import org.junit.Test;

/**
 * Test class for AsyncAgent demonstrating asynchronous agent invocation patterns.
 *
 * <p>This test demonstrates:
 *
 * <ul>
 *   <li>agent.initiate: Starting agent execution asynchronously
 *   <li>agent.result: Getting results from async execution
 *   <li>AgentInvoke handles for tracking execution
 * </ul>
 */
public class AsyncAgentTest {

  @Test
  public void testAsyncAgent() throws Exception {
    // Tests basic async agent invocation and result retrieval
    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Deploy the agent module
      AsyncAgent.AsyncAgentModule module = new AsyncAgent.AsyncAgentModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("AsyncAgent");

      // Test single async invocation
      AgentInvoke invoke = agent.initiate("TestTask");
      assertNotNull("AgentInvoke should not be null", invoke);

      String result = (String) agent.result(invoke);
      assertNotNull("Result should not be null", result);
      assertEquals("Task 'TestTask' completed successfully", result);
    }
  }
}
